pkg/helpers/status.go

```diff
@@ -0,0 +1,228 @@
+	instance.DeployFailed("workflow in failed state, starting rollback...")
+
+	return helper.patchStatus(ctx, instance)
+}
+
+// RollingBack sets the meta.ReadyCondition to 'True' and
+// meta.RollingBack reason and message
+func (helper *StatusHelper) RollingBack(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
+	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionTrue, meta.FailedReason, "workflow failed because of helmreleases during upgrade, rolling back...")
+	meta.SetResourceCondition(instance, meta.DeployCondition, metav1.ConditionTrue, meta.RollingBackReason, "rolling back because of failed helm releases...")
+
+	return helper.patchStatus(ctx, instance)
+}
+
+// Progressing resets the conditions of the ApplicationGroup to
+// metav1.Condition of type meta.ReadyCondition with status 'Unknown' and
+// meta.StartingReason reason and message.
+func (helper *StatusHelper) Progressing(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
+	instance.Status.Conditions = []metav1.Condition{}
+	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionUnknown, meta.ProgressingReason, "workflow is reconciling...")
+	meta.SetResourceCondition(instance, meta.DeployCondition, metav1.ConditionUnknown, meta.ProgressingReason, "application group is reconciling...")
+
+	return helper.patchStatus(ctx, instance)
+}
+
+// MarkReversing sets the meta.ReadyCondition to 'False', with the given
+// meta.Reversing reason and message
+func (helper *StatusHelper) MarkReversing(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
+	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionFalse, meta.ReversingReason, "application group is reversing...")
+	return helper.patchStatus(ctx, instance)
+}
+
+func InitAppStatus(appGroup *v1alpha1.ApplicationGroup) {
+	// Initialize the Status fields if not already setup
+	appGroup.Status.Applications = make([]v1alpha1.ApplicationStatus, 0, len(appGroup.Spec.Applications))
+	for _, app := range appGroup.Spec.Applications {
+		status := v1alpha1.ApplicationStatus{
+			Name:        app.Name,
+			ChartStatus: v1alpha1.ChartStatus{Version: app.Spec.Chart.Version},
+			Subcharts:   make(map[string]v1alpha1.ChartStatus),
+		}
+		appGroup.Status.Applications = append(appGroup.Status.Applications, status)
+	}
+}
+
+func (helper *StatusHelper) getWorkflowStatus(ctx context.Context, appGroupName string) (v1alpha12.NodePhase, error) {
+	wfs := v1alpha12.WorkflowList{}
+	listOption := client.MatchingLabels{
+		workflow.OwnershipLabel: appGroupName,
+		workflow.HeritageLabel:  workflow.Project,
+	}
+	err := helper.List(ctx, &wfs, listOption)
+	if err != nil {
+		helper.Error(err, "failed to find generated workflow instance")
+		return "", err
+	}
+	if wfs.Items.Len() == 0 {
+		err = fmt.Errorf("no associated workflow found")
+		helper.Error(err, "no associated workflow found")
+		return "", err
+	}
+	return wfs.Items[0].Status.Phase, nil
```

Comment:
Let's collect more than just the phase so that we can reflect the actual error reason when things fail. This would possibly address #289 as well